src/util/backend.js

```diff
@@ -0,0 +1,313 @@
+    if (!existing) {
+      this.cache[transferWithInfo.transfer.id] = transferWithInfo
+      existing = (this.store && (await this.store.get(this.key + ':tl:transfer:' + transfer.id)))
+      if (existing) {
+        delete this.cache[transferWithInfo.transfer.id]
+        return
+      }
+    }
+
+    if (existing) {
+      if (!deepEqual(existing.transfer, transferWithInfo.transfer)) {
+        throw new Error('transfer ' + JSON.stringify(transferWithInfo) +
+          ' matches the id of ' + JSON.stringify(existing) +
+          ' but not the contents.')
+      }
+      return
+    }
+
+    const balance = isIncoming ? 'balance_i' : 'balance_o'
+    const otherBalance = this[isIncoming ? 'balance_of' : 'balance_if']
+
+    const amount = transferWithInfo.transfer.amount
+    const isOver = isIncoming
+      ? (n) => n.sub(otherBalance).gt(this.maximum)
+      : (n) => n.sub(otherBalance).gt(this.minimum.neg())
+
+    if (isOver(this[balance].add(amount))) {
+      throw new Error(balance + ' exceeds greatest allowed value after: ' +
+        JSON.stringify(transferWithInfo))
+    }
+
+    this[balance] = this[balance].add(amount)
+    this.cache[transferWithInfo.transfer.id] = transferWithInfo
+
+    if (this.store) {
+      this.writeQueue = this.writeQueue
+        .then(() => {
+          return this.store.put(this.key + ':tl:transfer:' + transfer.id,
+            JSON.stringify(transferWithInfo))
+        })
+
+      await this.writeQueue
+    }
+  }
+
+  async fulfill (transferId, fulfillment) {
+    await this.connect()
+    // TODO: errors
+    // - what if a transfer is already fulfilled?
+    // - what if transfer doesn't exist?
+    // - should the fulfillment be validated?
+    // - what if the transfer is rejected?
+    const transferWithInfo = this.cache[transferId]
+    const isIncoming = transferWithInfo.isIncoming
+    const balance = isIncoming ? 'balance_if' : 'balance_of'
+
+    if (transferWithInfo.state === 'prepared') {
+      this[balance] = this[balance].add(transferWithInfo.transfer.amount)
+    }
+
+    // TODO: should the failure state be rejected, like FBL API?
+    if (transferWithInfo.state === 'cancelled') {
+      throw new Error(transferId + ' cannot be fulfilled because it is rejected: ' +
+        JSON.stringify(transferWithInfo))
+    }
+
+    transferWithInfo.state = 'fulfilled'
+    transferWithInfo.fulfillment = fulfillment
+    delete this.cache[transferId]
+
+    if (this.store) {
+      this.writeQueue = this.writeQueue
+        .then(() => {
+          return this.store.put(this.key + ':tl:transfer:' + transferWithInfo.transfer.id,
+            JSON.stringify(transferWithInfo))
+        }).then(() => {
+          const balanceKey = isIncoming ? ':tl:balance:if' : ':tl:balance:of'
+          return this.store.put(this.key + balanceKey, this[balance].toString())
+        })
+
+      await this.writeQueue
+    }
+  }
+
+  // TODO: should there be some kind of rejectionReason field? it's useful in FBL.
```

Comment:
rejectionReason is still included in the plugin, but it's not kept in the store.